cql3/query_processor.cc

```diff
@@ -1018,16 +1026,28 @@ query_processor::execute_schema_statement(const statements::schema_altering_stat
 
     cql3::cql_warnings_vec warnings;
 
+    auto request_id = guard->new_group0_state_id();
+    const_cast<statements::schema_altering_statement&>(stmt).global_req_id = fmt::to_string(request_id); // TODO: either that or extend prepare_schema_mutations with guard's req id
+
     auto [ret, m, cql_warnings] = co_await stmt.prepare_schema_mutations(*this, guard->write_timestamp());
     warnings = std::move(cql_warnings);
 
+    ce = std::move(ret);
     if (!m.empty()) {
         auto description = format("CQL DDL statement: \"{}\"", stmt.raw_cql_statement);
-        co_await remote_.get().mm.announce(std::move(m), std::move(*guard), description);
+        if (ce && ce->target == cql_transport::event::schema_change::target_type::TABLET_KEYSPACE) {
+            co_await remote_.get().mm.announce<service::topology_change>(std::move(m), std::move(*guard), description);
+            // TODO: eliminate timeout from alter ks statement
+            auto error = co_await remote_.get().ss.wait_for_topology_request_completion(request_id);
```

Comment:
@tgrabiec I'm thinking how to wait for tablets load balancer to complete here. Currently I'm only waiting for mutations to happen, which doesn't imply that they will be executed before we return response to the user.
I can see that e.g. `storage_service::transit_tablet` waits for topo_sm.event and checks if a certain tablet is in transition, but would always some tablet be in a transition when executing ALTER KS? E.g. some tablets can be added when increasing RF, but we can't tell which ones and which ones to wait for. OTOH are to-be-deleted tablets going through some transition? Other than these 2 cases (adding/removing tablets), other tablets won't go through any transition at all, so we cannot wait on them in any way as well. 
If waiting for specific tablets transition to complete is not an option, then I thought maybe waiting for some other topo.sm state could work, but here I also don't have any good ideas. Waiting for global rf_change req to complete is already implemented above by:
```
            auto error = co_await remote_.get().ss.wait_for_topology_request_completion(request_id);
```
Is tablets load balancer's state connected to topo.sm somehow maybe? So that I could wait for balancer to complete through topo.sm? But even if, maybe I wouldn't be able to wait exactly for the event I initiated, because balancer could be still processing an older event and if I was to wait for the older event to complete, I'd end up in the exact same spot as now. Or (big maybe) balancer could already be processing some next event, and waiting for the next event would only prolong generally waiting until alter completes?

Reply:
> I can see that e.g. storage_service::transit_tablet waits for topo_sm.event and checks if a certain tablet is in transition, but would always some tablet be in a transition when executing ALTER KS?

The proper way to wait depends on how background rebuilding will be carried out. If we put all tablets in transition state inside the request, then it's enough to wait for topology to not be busy, as that implies that all tablets finished rebuilding, because: (1) already started transitions cannot be canceled (2) topology exits the "tablet_migration" track only when there are no active transitions. 

`_topology_state_machine._topology.is_busy()`.

Reply:
\cc @xemul

Reply:
OK, thanks, is that what you had in mind - https://github.com/scylladb/scylladb/pull/16723/commits/3593109a293a613e1bdb317d89c77456d82ab9d7 ?